project/src/com/google/daggerqueryui/scripts/analyze_input_field_content.js

```diff
@@ -0,0 +1,103 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+$(function () {
+  $.supportedQueries = new Map([
+    ['deps', 1],
+    ['allpaths', 2],
+    ['somepath', 2],
+  ]);
+
+  /**
+   * Validates the request name to see if it is supported or not.
+   *
+   * @param {string} queryName the name of the query to check
+   * @return {boolean} a value that determines whether the request is supported or not
+   */
+  $.fn.validateQueryName = function (queryName) {
+    return $.supportedQueries.has(queryName);
+  };
+
+  /**
+   * Validates the number of query parameters.
+   *
+   * @param {Array<string>} query a full query passed by user with unique name and parameters
+   * @return {boolean} a value that determines whether the number of parameters is correct or not
+   */
+  $.fn.validateParameters = function (query) {
+    return $.supportedQueries.get(query[0]) === query.length - 1;
+  };
+});
+
+$(function () {
+  /**
+   * Marks input field as invalid and shows a container with error message.
+   * @param {string} errorMessage a error message which will be shown under the input field
+   */
+  $.fn.markInputFieldAsInvalid = function (errorMessage) {
+    const inputField = document.getElementById("query-input");
+    const errorContainer = document.getElementById("error-message");
+
+    inputField.classList.remove("is-valid");
+    inputField.classList.add("is-invalid");
+    errorContainer.classList.remove("hidden");
+    errorContainer.innerHTML = errorMessage;
```

Comment:
IIUC there's a risk of an XSS injection attack here (however one caused by and only affecting the user).

From the [gif](https://user-images.githubusercontent.com/19249980/92711385-2bc61300-f361-11ea-9682-8687c2840182.gif) you provided it looks like the input is directly relayed back to the user from the HTTP request and set as HTML in the error-message div. A user could potentially enter HTML markup (or a script tag) which would then be rendered as HTML markup in the div, due to the usage of the `innerHTML` attribute.

As you're expecting a text response from the server (and not HTML code) using `el.innerText` is the better option. If you plan on returning HTML in the future such as text encased in <b> tags, take a look into [goog.dom.safe](https://google.github.io/closure-library/api/goog.dom.safe.html), this would still require sanitising user provided input though.

See http://go/xss#dom-based-xss and http://go/xss-antipatterns#example-21-using-untrusted-input-to-set-innerhtml-in-javascript

Reply:
or use the jquery `.text(text)` setter which automatically escapes the contents

Reply:
That's very interesting, thanks. At this point `.text(...)` looks attractive because all the logic for representing errors can be simplified to a [single line](https://github.com/googleinterns/dagger-query/blob/3ce235a5311d3e3225bad54f6acabd1d71add380/project/src/com/google/daggerqueryui/scripts/analyze_input_field_content.js#L50).